### DISCUSSION TITLE
Reorder keys in a specific order

### DISCUSSION THREAD

**Discussion by sytone** [mikefarah/yq#1778] [Q&A]
I know I can sort keys but is there a way to reorder the keys in a specific way?

I am working with markdown files with yaml front matter and was wondering if I can process the file and provide a specific ordering of the keys in the header. Mine are a bit more complect but the below gives a idea of what I want.

From this:
```
---
area: yard
version: 3
notetype: ideas
tags:
 - tag1
 - tag2
next-review: 2023-11-01
---
```
to this:
```
---
notetype: ideas
version: 3
tags:
 - tag1
 - tag2
area: yard
next-review: 2023-11-01
---
```

**Comment by mikefarah** [comment]
Yep, you can use the `pick` operator: https://mikefarah.gitbook.io/yq/operators/pick#pick-keys-from-map

**Reply by Aubermean** [reply]
Hi @mikefarah, how can I not skip non existent keys?

**Reply by Aubermean** [reply]
At the moment this is my solution:

```bash
yq eval-all '. | 
pick(["apiVersion", "kind", "metadata", "spec"])' kustomization.yaml | 
yq eval-all 'select(fileIndex == 0) * select(fileIndex == 1)' - kustomization.yaml | 
awk '!/^#/ || !seen[$0]++
```

The last awk is to prevent duplicated comment lines remaining after the merge. Maybe there's a cleaner way? Thanks!

**Reply by mikefarah** [reply]
You can do it by creating an array of the keys, with dupes, then de-dupe using unique:

```
yq 'pick( (["notetype"] + keys) | unique)' examples/data1.yaml
```

**Reply by salim-b** [reply]
Addendum: To leave the order of the not-explicitly-picked keys intact, use map [merge operator `*`](https://mikefarah.gitbook.io/yq/operators/multiply-merge#merge-objects-together-returning-merged-result-only) like this, optionally with the [`n` modifier](https://mikefarah.gitbook.io/yq/operators/multiply-merge#merge-only-new-fields) (tiny little bit more efficient than without; same result):

```sh
yq --front-matter=process '. as $data | pick(["apiVersion", "kind", "metadata", "spec"]) | . *n $data' kustomization.yaml
```
